File: solutions/javascript/captains-log/1/captains-log.js
// @ts-check

/**
 * Generates a random starship registry number.
 * Registry numbers start with "NCC-" and then use a number from 1000 to 9999 (both inclusive).
 *
 * @returns {string} The generated registry number (e.g., "NCC-1947").
 */
export function randomShipRegistryNumber() {
  // Math.random() returns a float between 0 (inclusive) and 1 (exclusive).
  // To get a number in a range [min, max] (inclusive of both), use:
  // Math.floor(Math.random() * (max - min + 1)) + min;
  const min = 1000;
  const max = 9999;
  const randomNumber = Math.floor(Math.random() * (max - min + 1)) + min;
  return `NCC-${randomNumber}`;
}

/**
 * Generates a random stardate.
 * A stardate is a floating point number between 41000.0 (inclusive) and 42000.0 (exclusive).
 *
 * @returns {number} A stardate (e.g., 41458.15721310934).
 */
export function randomStardate() {
  // To get a number in a range [min, max) (inclusive of min, exclusive of max), use:
  // Math.random() * (max - min) + min;
  const minStardate = 41000.0;
  const maxStardate = 42000.0; // Exclusive
  return Math.random() * (maxStardate - minStardate) + minStardate;
}

/**
 * Generates a random planet class.
 * Possible planetary classes are: D, H, J, K, L, M, N, R, T, and Y.
 *
 * @returns {string} A one-letter planet class (e.g., "K").
 */
export function randomPlanetClass() {
  const planetClasses = ['D', 'H', 'J', 'K', 'L', 'M', 'N', 'R', 'T', 'Y'];

  // To pick a random element from an array, generate a random index:
  // Math.floor(Math.random() * array.length);
  const randomIndex = Math.floor(Math.random() * planetClasses.length);

  return planetClasses[randomIndex];
}

File: solutions/java/captains-log/1/src/main/java/CaptainsLog.java
import java.util.Random;

class CaptainsLog {

    private static final char[] PLANET_CLASSES = new char[]{'D', 'H', 'J', 'K', 'L', 'M', 'N', 'R', 'T', 'Y'};

    private Random random;

    CaptainsLog(Random random) {
        this.random = random;
    }

    // Task 1: Generate a random planet class
    char randomPlanetClass() {
        int index = random.nextInt(PLANET_CLASSES.length);
        return PLANET_CLASSES[index];
    }

    // Task 2: Generate a random starship registry number (NCC-1000 to NCC-9999)
    String randomShipRegistryNumber() {
        int number = 1000 + random.nextInt(9000); // 9999 - 1000 + 1 = 9000
        return "NCC-" + number;
    }

    // Task 3: Generate a random stardate (41000.0 inclusive to 42000.0 exclusive)
    double randomStardate() {
        return 41000.0 + 1000.0 * random.nextDouble();
    }
}
